OrbitVulkanLayer/DispatchTableTest.cpp

```diff
@@ -0,0 +1,803 @@
+
+  VkQueue queue = {};
+  VkFence fence = {};
+  VkResult result = dispatch_table.QueueSubmit(device)(queue, 0, nullptr, fence);
+  EXPECT_EQ(result, VK_SUCCESS);
+}
+
+TEST(DispatchTable, CanCallQueuePresentKHR) {
+  // We can not create an actual VkInstance, but the first bytes of an any dispatchable type in
+  // Vulkan will be a pointer to a dispatch table. This characteristic will be used by our dispatch
+  // table wrapper, so we need to mimic it.
+  VkLayerDispatchTable some_dispatch_table = {};
+  auto device = absl::bit_cast<VkDevice>(&some_dispatch_table);
+
+  PFN_vkGetDeviceProcAddr next_get_device_proc_addr_function =
+      +[](VkDevice /*device*/, const char* name) -> PFN_vkVoidFunction {
+    if (strcmp(name, "vkQueuePresentKHR") == 0) {
+      PFN_vkQueuePresentKHR function = +[](VkQueue /*queue*/, const VkPresentInfoKHR *
+                                           /*present_info*/) -> VkResult { return VK_SUCCESS; };
+      return absl::bit_cast<PFN_vkVoidFunction>(function);
+    }
+    return nullptr;
+  };
+
+  DispatchTable dispatch_table = {};
+  dispatch_table.CreateDeviceDispatchTable(device, next_get_device_proc_addr_function);
+
+  VkQueue queue = {};
+  VkResult result = dispatch_table.QueuePresentKHR(device)(queue, nullptr);
+  EXPECT_EQ(result, VK_SUCCESS);
+}
+
+TEST(DispatchTable, CanCallCreateQueryPool) {
+  // We can not create an actual VkInstance, but the first bytes of an any dispatchable type in
+  // Vulkan will be a pointer to a dispatch table. This characteristic will be used by our dispatch
+  // table wrapper, so we need to mimic it.
+  VkLayerDispatchTable some_dispatch_table = {};
+  auto device = absl::bit_cast<VkDevice>(&some_dispatch_table);
+
+  PFN_vkGetDeviceProcAddr next_get_device_proc_addr_function =
+      +[](VkDevice /*device*/, const char* name) -> PFN_vkVoidFunction {
+    if (strcmp(name, "vkCreateQueryPool") == 0) {
+      PFN_vkCreateQueryPool function =
+          +[](VkDevice /*device*/, const VkQueryPoolCreateInfo* /*create_info*/,
+              const VkAllocationCallbacks* /*allocator*/, VkQueryPool *
+              /*query_pool*/) -> VkResult { return VK_SUCCESS; };
+      return absl::bit_cast<PFN_vkVoidFunction>(function);
+    }
+    return nullptr;
+  };
+
+  DispatchTable dispatch_table = {};
+  dispatch_table.CreateDeviceDispatchTable(device, next_get_device_proc_addr_function);
+
+  VkResult result = dispatch_table.CreateQueryPool(device)(device, nullptr, nullptr, nullptr);
+  EXPECT_EQ(result, VK_SUCCESS);
+}
+
+TEST(DispatchTable, CanCallResetQueryPoolEXT) {
+  // We can not create an actual VkInstance, but the first bytes of an any dispatchable type in
+  // Vulkan will be a pointer to a dispatch table. This characteristic will be used by our dispatch
+  // table wrapper, so we need to mimic it.
+  VkLayerDispatchTable some_dispatch_table = {};
+  auto device = absl::bit_cast<VkDevice>(&some_dispatch_table);
+
+  static bool was_called = false;
```

Comment:
Why is this static? You don't want to preserve the state across different runs, right?

Reply:
It needs to be static, such that I can use the lambda as function pointer. However, you are right, I should at least set it to false after the check again.

Reply:
I've the static bool to false at the end of each test.

Reply:
Can we assume that each test is run only once per execution of the test binary?

Reply:
That's a good question. So far, yes. Especially, as we would only have an issue as the same test gets executed from the same process in parallel.

Reply:
Ok, that was to say that maybe it's a bit weird to reset `was_called` if you don't expect this function to be called again. But on the other hand I guess it's theoretically more precise.

Reply:
Sorry if this is not on point, but one can repeat tests multiple times with --gtest_repeat=10 (or whatever number you want), which is useful when debugging a flaky test. This does not run anything in parallel, though.

Reply:
Right, I was also thinking of this. In this case, the reset at the end should be the right thing to do.

Reply:
Maybe you can add a comment in the first test that does this explaining why this needs to be static (basically saying what you replied to Daniel).

Reply:
This might be a good case for using a test fixture (TEST_F) as it seems you need to reset state after each test. You can also put most of the common setup code into that.

Reply:
Regarding the TEST_F, if I get it correctly, then we would have one static bool for all the tests, rather then one per test, right? Are we sure, that no tests are executed in parallel?

Reply:
Anyways added a comment to the first usage.

Reply:
Depends on whether you create a test fixture for all tests or one for each test, though the latter seems a bit too much overhead and boilerplate. I guess best to ignore this idea :)

Reply:
Agree, I would have been nice, but I guess it would also be too much overhead. Let's keep it as it for now.